hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientPNCounterProxy.java

```diff
@@ -0,0 +1,399 @@
+            return invokeInternal(request, excludedAddresses, e, newTarget);
+        }
+    }
+
+    /**
+     * Returns the target on which this proxy should invoke a CRDT operation.
+     * On first invocation of this method, the method will choose a target
+     * address and return that address on future invocations. Replicas with
+     * addresses contained in the {@code excludedAddresses} list are excluded
+     * and if the chosen replica is in this list, a new replica is chosen and
+     * returned on future invocations.
+     * The method may return {@code null} if there are no viable target addresses.
+     *
+     * @param excludedAddresses the addresses to exclude from returning from
+     *                          this method
+     * @return a CRDT replica address or {@code null} if there are no viable
+     * addresses
+     */
+    private Address getCRDTOperationTarget(ArrayList<Address> excludedAddresses) {
+        if (currentTargetReplicaAddress != null &&
+                (excludedAddresses == null || !excludedAddresses.contains(currentTargetReplicaAddress))) {
+            return currentTargetReplicaAddress;
+        }
+
+        synchronized (targetSelectionMutex) {
+            if (currentTargetReplicaAddress == null
+                    || (excludedAddresses != null && excludedAddresses.contains(currentTargetReplicaAddress))) {
+                currentTargetReplicaAddress = chooseTargetReplica(excludedAddresses);
+            }
+        }
+        return currentTargetReplicaAddress;
+    }
+
+    /**
+     * Chooses and returns a CRDT replica address. Replicas with addresses
+     * contained in the {@code excludedAddresses} list are excluded and the
+     * method chooses randomly between the collection of viable target addresses.
+     * <p>
+     * The method may return {@code null} if there are no viable addresses.
+     *
+     * @param excludedAddresses the addresses to exclude when choosing a replica address
+     * @return a CRDT replica address or {@code null} if there are no viable addresses
+     */
+    private Address chooseTargetReplica(ArrayList<Address> excludedAddresses) {
+        final List<Address> replicaAddresses = getReplicaAddresses(excludedAddresses);
+        if (replicaAddresses.isEmpty()) {
+            return null;
+        }
+        final int randomReplicaIndex = ThreadLocalRandomProvider.get().nextInt(replicaAddresses.size());
+        return replicaAddresses.get(randomReplicaIndex);
+    }
+
+    /**
+     * Returns the addresses of the CRDT replicas from the current state of the
+     * local membership list. Addresses contained in the {@code excludedAddresses}
+     * collection are excluded.
+     *
+     * @param excludedAddresses the addresses to exclude
+     * @return list of possible CRDT replica addresses
+     */
+    private List<Address> getReplicaAddresses(ArrayList<Address> excludedAddresses) {
```

Comment:
I wanted to avoid creating a new list for every invocation of every get or update method, even in the case where we always go to the same data member. WDYT?

Reply:
maybe we can create a static empty list and pass it initially instead of a null value.

Reply:
Thanks, will do.